### ISSUE TITLE
set aio_threads not working 

### ISSUE THREAD

**Issue by Joshua-Wang** [arvidn/libtorrent#3704]
**Please provide the following information**

libtorrent version (or branch):
1.1.4

platform/architecture:

compiler and compiler version:

please describe what symptom you see, what you would expect to see instead and
how to reproduce it.

I have set aio_threads to 64 for better performance. But it doesn't seem to work. Following shows that this process only have 10 threads. I have tried higher versions of libtorrent, but still it doesn't work.
By the way I'm using python bindlings.  Any help will be greatly appreciated.

![image](https://user-images.githubusercontent.com/6940489/54178595-5c5d9000-44d1-11e9-9d7f-869568feb7f6.png)

**Comment by arvidn** [comment]
iirc, new threads are spawned on demand. If there's a new disk I/O job and all existing threads are busy, a new one is created. So, if you don't put enough load on the disk subsystem, it may not create all threads

**Comment by arvidn** [comment]
there's only a single network thread in libtorrent, unfortunately it's quite complicated to split it up over multiple threads as they heavily depend on shared state in the torrent object and session object.

You should probably make sure you understand which threads are using the CPU and what they are doing though. It may not be the network thread that's the bottleneck. Are you running bittorrent over SSL by any chance?

**Comment by arvidn** [comment]
are you uploading or downloading?

**Comment by Joshua-Wang** [comment]
I am using libtorrent both for uploading and downloading. There is one server for making torrent and uploading, while about 80 servers downloading from it.  I guess SSL is not used here.

**Comment by Joshua-Wang** [comment]
I'll try parallelize the stage of making torrent. Since the source file is huge, about 50g bytes, split it into several files, and then make torrent and upload concurrently might help.

**Comment by Joshua-Wang** [comment]
> there's only a single network thread in libtorrent, unfortunately it's quite complicated to split it up over multiple threads as they heavily depend on shared state in the torrent object and session object.
> 
> @arvidn  I've parallelized the stage of making torrent and it helped a lot. So can we create several lt.session() object, and add_torrent seperately in different threads, could it speed things up ?

**Comment by arvidn** [comment]
yes, that would give you multiple network threads. I would be interested in understanding exactly what your bottleneck is. Normally the network thread can push at least ~1 GB/s before becoming the bottleneck (on a normal laptop/desktop). It doesn't sound like you're using SSL (which adds a lot of expensive crypto calculation to the network thread) nor run on an embedded system (where the single thread performance probably can't get to 1 GB/s).

It sounds like the other 80 downloaders also upload to each other, which would further reduce the pressure on a single machine.

Did you disable uTP? if not, try setting `enable_outgoing_utp` and `enable_incoming_utp` to false on the sessions. Since uTP sits on top of UDP, and UDP is typically more expensive, that may cost you.

**Comment by arvidn** [comment]
when you say you parallelized "the stage of making torrent", do you literally mean computing the piece hashes for the content? Are you doing that by using more threads in the `disk_io_thread` object? I realise the number of threads for `set_piece_hashes()` is not configurable right now, but it shouldn't be too hard to pass in an argument, or a `settings_pack` object perhaps.

**Comment by Joshua-Wang** [comment]
I have already set enable_outgoing_utp and enable_incoming_utp to false . Now the bottleneck I saw is lt.session().add_torrent(), since I will produce about 20 files once, and than make_torrent for each file concurrently , and **then use session object to add_torrent() serially. (This takes quit some time, because it has to calculate pieces and compare with the torrents )** . So I would like to use several session object in different threads.

**Comment by arvidn** [comment]
if you just created the torrents, by hashing the files, there really isn't any reason to hash them again when adding them to the session. there are two main options, either add the torrent(s) in `seed_mode` (that's a flag on `add_torrent_params`). That's the easiest, it will assume all pieces are there and defer checking the hash.

Alternatively, you can construct a resume data indicating that you have all pieces, which will avoid the hash checking entirely. This is complicated in libtorrent-1.1 but quite simple in libtorrent-1.2. in 1.1 you actually have to construct a bencoded dictionary and fill it in with knowledge about the resume data format. in 1.2 there's another field in the `add_torrent_params` object.

You should do one of those. However, checking hashes should already be done in parallel. one of of every 4 aio thread is dedicated to hashing (this is to not starve out actual disk-I/O).

if you set `aio_threads` to 32, there will be 8 threads hashing pieces. Is that not what you see?

**Comment by Joshua-Wang** [comment]
That helped a lot, Thanks.

**Comment by Joshua-Wang** [comment]
add the torrent(s) in seed_mode (that's a flag on add_torrent_params).

This helped a lot , Thanks.
